### PR TITLE
Fix doc on GCM API

### LINKED ISSUE #9599: Wrong usage of mbedtls_gcm_crypt_and_tag() during TLS message decryption
mbedTLS version: 3.6.0
Platform: ARM STM32

While hunting a bug we noticed that the TLS buffer decryption does not follow the calling convention of mbedtls_gcm_crypt_and_tag() and mbedtls_gcm_auth_decrypt() (which calls mbedtls_gcm_crypt_and_tag() with the parameter MBEDTLS_GCM_DECRYPT).

Both function state that:
> For decryption, the output buffer cannot be the same as
> input buffer. If the buffers overlap, the output buffer
> must trail at least 8 Bytes behind the input buffer.

mbedtls_ssl_decrypt_buf() calls mbedtls_cipher_auth_decrypt_ext() in line 1638 with 'data' as input and output buffer.
These pointers are passed down to the aforementioned functions mostly unaltered as 'input' and 'output', but end up as the same pointer in mbedtls_gcm_crypt_and_tag() (as confirmed during debugging).

Please have a look at the attached StackTrace for details.
image

### PR DESCRIPTION
**Pull request by ivq**
## Description

Fixes #9599

The doc of some GCM API is out-dated.
The replaced content is true for initial GCM implementation but not for now.

### When did the wrong doc start to exist?
It originated from commit ca4ab491585b432d7db0266aa6fbcbc6813eda00.
### When did the old doc become wrong?
 Commit 09d67258a2a92831c9d49f9f845c2691ea5bb613.

## PR checklist

- [x] **changelog** not required
- [x] **3.6 backport** provided: #9762.
- [x] **2.28 backport** provided: #9774.
- [x] **tests**  not required

### CODE REVIEW

File: include/mbedtls/gcm.h [hunk 1]
```diff
@@ -298,9 +281,9 @@ int mbedtls_gcm_update_ad(mbedtls_gcm_context *ctx,
  *                      is greater than zero, this must be a writable buffer of
  *                      of at least \p output_size bytes.
  * \param output_size   The size of the output buffer in bytes.
- *                      See the function description regarding the output size.
  * \param output_length On success, \p *output_length contains the actual
- *                      length of the output written in \p output.
+ *                      length of the output written in \p output, which equals
```
Comment (Ryan-Everett-arm) [inline include/mbedtls/gcm.h:285]: This should say: "On success, \p *output_length contains the actual length of the output written in \p output, which is equal to \p input_length."

**Review by Ryan-Everett-arm — CHANGES_REQUESTED** [review]
Hi Chien, just one small grammatical issue, other than that it looks good.

File: include/mbedtls/gcm.h (hunk 1 above)
Comment (ivq) [reply include/mbedtls/gcm.h]: Thanks. It's fixed.

**Review by Ryan-Everett-arm — DISMISSED** [review]

**Comment by ivq** [comment]
We still need one more reviewer to make progress.

File: include/mbedtls/gcm.h [hunk 2]
```diff
@@ -269,25 +268,9 @@ int mbedtls_gcm_update_ad(mbedtls_gcm_context *ctx,
...
- *                    of 16-byte blocks. If the total input length so far
- *                    (not including associated data) is 16 \* *B* + *A*
- *                    with *A* < 16 then the total output length is 16 \* *B*.
- *
- *                  In particular:
- *                  - It is always correct to call this function with
- *                    \p output_size >= \p input_length + 15.
- *                  - If \p input_length is a multiple of 16 for all the calls
- *                    to this function during an operation, then it is
- *                    correct to use \p output_size = \p input_length.
```
[the remark this answers is not in the document]
Comment (ivq) [reply include/mbedtls/gcm.h:286]: That was a mistake. The deleted documentation is also for potential alternative implementations, added in #4342. Fixed now.

File: include/mbedtls/gcm.h [hunk 3]
```diff
@@ -113,10 +113,9 @@ int mbedtls_gcm_setkey(mbedtls_gcm_context *ctx,
 /**
  * \brief           This function performs GCM encryption or decryption of a buffer.
  *
- * \note            For encryption, the output buffer can be the same as the
- *                  input buffer. For decryption, the output buffer cannot be
- *                  the same as input buffer. If the buffers overlap, the output
- *                  buffer must trail at least 8 Bytes behind the input buffer.
+ * \note            The output buffer \p output can be the same as the input
+ *                  buffer \p input. If \p output is greater than \p input, they
+ *                  cannot overlap.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/gcm.h:118]: So this is preexisting, but we aren't testing anything regarding overlap for the GCM API. I'm uncomfortable making a stronger promise without adding tests to ensure that we won't break that stronger promise.

**Review by gilles-peskine-arm — DISMISSED** [review]
While it would be nice to make the documentation more precise, I am uncomfortable with relaxing the API without having a test for the stronger promise. So I'm sorry but I'm requesting to either not relax the API promise, or add tests.

**Comment by gilles-peskine-arm** [comment]
Furthermore, there's another reason why we can't relax the API in LTS branches: it invalidates implementations of `MBEDTLS_GCM_ALT` that only obey the documented constraint.

**Comment by tom-cosgrove-arm** [comment]
But [code within Mbed TLS itself calls the API with input == output](https://github.com/Mbed-TLS/mbedtls/issues/9599), so any _ALT implementation won't work with the TLS stack

**Comment by gilles-peskine-arm** [comment]
@tom-cosgrove-arm In my experience, ALT implementations are not always tested against the TLS stack before they're published. Also the behavior with respect to overlap can sometimes depend on compiler optimizations. So we need to tread carefully.

**Comment by davidhorstmann-arm** [comment]
> Furthermore, there's another reason why we can't relax the API in LTS branches: it invalidates implementations of `MBEDTLS_GCM_ALT` that only obey the documented constraint.

I would argue that we can work round this. If we have a test for our implementation we can say:

> When `MBEDTLS_GCM_ALT` is disabled, the output buffer `output` can overlap with the input buffer `input`. If `output` is greater than `input`, they cannot overlap.
>
> When `MBEDTLS_GCM_ALT` is enabled and `output` overlaps with `input`, `output` must trail `input` by at least 8 bytes.

**Comment by davidhorstmann-arm** [comment]
@ivq Sorry that this PR has been forgotten about. Do you have time to work on it or would you be okay with one of us taking it over to get it through?

**Comment by ivq** [comment]
> @ivq Sorry that this PR has been forgotten about. Do you have time to work on it or would you be okay with one of us taking it over to get it through?

It's fine that you take it over. I do not have time to work on it now.

**Comment by Harry-Ramsey** [comment]
I have added some tests for encryption/decryption using GCM where data is copied into a buffer that is used for both input and output. The test cases are simply copies of existing test cases where they used individual buffers for input and output. 

Hopefully upon review this documentation change and testing improvement can be merged into development.

File: tf-psa-crypto/tests/suites/test_suite_gcm.function [hunk 4]
```diff
@@ -605,6 +605,146 @@ exit:
...
+
+    BLOCK_CIPHER_PSA_INIT();
+    mbedtls_gcm_init(&ctx);
+
+    /* GCM includes padding and therefore input length can be shorter than the output length
+     * Therefore we must ensure we round up to the nearest 128-bits/16-bytes.
+     */
+    buffer_len = src_str->len;
+    if (buffer_len % 16 != 0) {
+      buffer_len += (16 - (buffer_len % 16));
```
Comment (ivq) [inline tf-psa-crypto/tests/suites/test_suite_gcm.function:631]: Too small indent

File: tf-psa-crypto/tests/suites/test_suite_gcm.function [hunk 5]
```diff
@@ -605,6 +605,146 @@ exit:
...
```
Comment (ivq) [inline tf-psa-crypto/tests/suites/test_suite_gcm.function:696]: Too small indent

Comment (Harry-Ramsey) [reply tf-psa-crypto/tests/suites/test_suite_gcm.function]: Thanks for the spot.

File: tf-psa-crypto/tests/suites/test_suite_gcm.function [hunk 6]
```diff
@@ -605,6 +605,146 @@ exit:
...
+        goto exit;
+    }
+    memcpy(buffer, src_str->x, src_str->len);
+
+    memset(tag_output, 0x00, 16);
+
+    TEST_ASSERT(mbedtls_gcm_setkey(&ctx, cipher_id, key_str->x, key_str->len * 8) == init_result);
+    if (init_result == 0) {
+        TEST_ASSERT(mbedtls_gcm_crypt_and_tag(&ctx, MBEDTLS_GCM_ENCRYPT, src_str->len, iv_str->x,
+                                              iv_str->len, add_str->x, add_str->len, src_str->x,
```
Comment (davidhorstmann-arm) [inline tf-psa-crypto/tests/suites/test_suite_gcm.function:645]: You're still passing non-overlapping buffers here, as `src_str->x` is a different buffer to `buffer`. I believe this should be:
```suggestion
                                              iv_str->len, add_str->x, add_str->len, buffer,
```

File: tf-psa-crypto/tests/suites/test_suite_gcm.function [hunk 7]
```diff
@@ -605,6 +605,146 @@ exit:
...
+    mbedtls_gcm_init(&ctx);
+
+    /* GCM includes padding and therefore input length can be shorter than the output length
+     * Therefore we must ensure we round up to the nearest 128-bits/16-bytes.
+     */
+    buffer_len = src_str->len;
+    if (buffer_len % 16 != 0) {
+      buffer_len += (16 - (buffer_len % 16));
+    }
+    buffer = malloc(buffer_len);
```
Comment (davidhorstmann-arm) [inline tf-psa-crypto/tests/suites/test_suite_gcm.function:633]: Best to use `TEST_CALLOC()` here as it handles all the edge cases:
```suggestion
    buffer = TEST_CALLOC(buffer_len, 1);
```

File: tf-psa-crypto/tests/suites/test_suite_gcm.function [hunk 8]
```diff
@@ -605,6 +605,146 @@ exit:
...
```
Comment (davidhorstmann-arm) [inline tf-psa-crypto/tests/suites/test_suite_gcm.function:698]: As above

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
A few comments on the tests.

Otherwise, it would be good to specify in the docs that overlap is not necessarily supported by ALT implementations when `MBEDTLS_GCM_ALT` is enabled, as we have to allow for other people's implementations of GCM adhering only to the minimum documented requirements.

File: tf-psa-crypto/tests/suites/test_suite_gcm.function (hunk 6 above)
Comment (Harry-Ramsey) [reply tf-psa-crypto/tests/suites/test_suite_gcm.function]: Thanks, fixed now.

File: tf-psa-crypto/tests/suites/test_suite_gcm.function (hunk 7 above)
Comment (Harry-Ramsey) [reply tf-psa-crypto/tests/suites/test_suite_gcm.function]: Thanks, fixed now.

File: tf-psa-crypto/tests/suites/test_suite_gcm.function (hunk 8 above)
Comment (Harry-Ramsey) [reply tf-psa-crypto/tests/suites/test_suite_gcm.function]: Thanks, fixed now.

File: tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h [hunk 9]
```diff
@@ -173,9 +172,11 @@ int mbedtls_gcm_crypt_and_tag(mbedtls_gcm_context *ctx,
...
  *                  buffer.
  *
- * \note            For decryption, the output buffer cannot be the same as
- *                  input buffer. If the buffers overlap, the output buffer
- *                  must trail at least 8 Bytes behind the input buffer.
+ * \note            The output buffer \p output can be the same as the input
+ *                  buffer \p input. If \p output is greater than \p input, they
+ *                  cannot overlap. Implementations which require
+ *                  MBEDTLS_GCM_ALT to be enabled may not provide support
+ *                  overlapping buffers.
```
Comment (davidhorstmann-arm) [inline tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h:179]: Typo:
```suggestion
 *                  for overlapping buffers.
```

File: tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h [hunk 10]
```diff
@@ -281,9 +282,11 @@ int mbedtls_gcm_update_ad(mbedtls_gcm_context *ctx,
...
  *                    correct to use \p output_size = \p input_length.
  *
- * \note            For decryption, the output buffer cannot be the same as
- *                  input buffer. If the buffers overlap, the output buffer
- *                  must trail at least 8 Bytes behind the input buffer.
+ * \note            The output buffer \p output can be the same as the input
+ *                  buffer \p input. If \p output is greater than \p input, they
+ *                  cannot overlap. Implementations which require
+ *                  MBEDTLS_GCM_ALT to be enabled may not provide support
+ *                  overlapping buffers.
```
[the remark this answers is not in the document]
Comment (Harry-Ramsey) [reply tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h:289]: Completed, thanks.

File: tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (Harry-Ramsey) [reply tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h:179]: Completed, thanks.

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, thanks!

**Review by eleuzi01 — COMMENTED** [review]
The CI is unhappy

**Comment by eleuzi01** [comment]
> Hopefully CI passes now and this can be re-reviewed and merged.

Doesn't look very promising so far

**Review by eleuzi01 — DISMISSED** [review]
LGTM, thanks!

File: tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h [hunk 11]
```diff
@@ -285,8 +285,8 @@ int mbedtls_gcm_update_ad(mbedtls_gcm_context *ctx,
  * \note            The output buffer \p output can be the same as the input
  *                  buffer \p input. If \p output is greater than \p input, they
  *                  cannot overlap. Implementations which require
- *                  MBEDTLS_GCM_ALT to be enabled may not provide support for
- *                  overlapping buffers.
+ *                  MBEDTLS_GCM_ALT //no-check-names to be enabled may not
```
Comment (davidhorstmann-arm) [inline tf-psa-crypto/drivers/builtin/include/mbedtls/gcm.h:288]: Ah. There's a good reason we have `check_names.py` telling us about this.

I'm very sorry, I forgot that we've removed `MBEDTLS_GCM_ALT` in `development`. Would you be able to remove the note about `MBEDTLS_GCM_ALT` in this PR? (But not in the backports).

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, thanks!

**Review by eleuzi01 — APPROVED** [review]

**Pull request merged**